### PR TITLE
PROJ6+ + WKT2 support

### PR DESCRIPTION
**Pull request by metzm**
New GRASS locations now store the projection definition also in WKT (WKT2 with GDAL 3 / PROJ 6) and the spatial reference ID (SRID) if available. Two new files are (if possible) created for a new location: PERMANENT/PROJ_SRID and PERMANENT/PROJ_WKT. The file PERMANENT/PROJ_EPSG is no longer created and superseeded by PROJ_SRID which supports not only EPSG, but also other authorities. An existing file PERMANENT/PROJ_EPSG is read if PROJ_SRID does not exist.

For reprojections, the priority of spatial reference definitions is 1. SRID, 2. WKT, 3. PROJ4 style from GRASS definition

changed libraries
gis, proj

changed modules
g.proj, r.proj, v.proj, r.in.gdal, v.in.ogr, r.external, v.external, r.out.gdal, v.out.ogr

This PR also solves https://github.com/qgis/QGIS/issues/18596#issuecomment-753481210

testing
reprojection from/to a CRS with neither meters nor degrees as units might not convert coordinates to map units

use location nc_spm_08
EPSG:3358
NAD83(HARN) / North Carolina

CRS with feet as units
EPSG:2264
NAD83 / North Carolina (ftUS)
location nc_epsg_2264

CRS with different datum
EPSG:32617
UTM zone 17N
location nc_epsg_32617

reprojection with projection definitions from SRIDs

reproject nc_state from nc_spm_08 to location nc_epsg_2264 to location nc_epsg_32617 and back to nc_spm_08
v.proj location=nc_spm_08 mapset=PERMANENT input=nc_state output=nc_state_srid_1
v.proj location=nc_epsg_2264 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_srid_1
v.proj location=nc_epsg_32617 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_srid_1

reproject nc_state from nc_spm_08 to location nc_epsg_32617 to location nc_epsg_2264 and back to nc_spm_08
v.proj location=nc_spm_08 mapset=PERMANENT input=nc_state output=nc_state_srid_2
v.proj location=nc_epsg_32617 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_srid_2
v.proj location=nc_epsg_2264 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_srid_2

end results in nc_spm_08 should be identical (within floating point precision limits)

now rename all PROJ_EPSG, PROJ_SRID, and PROJ_WKT files in the three locations

reprojection with standard GRASS projection definitions

reproject nc_state from nc_spm_08 to location nc_epsg_2264 to location nc_epsg_32617 and back to nc_spm_08
v.proj location=nc_spm_08 mapset=PERMANENT input=nc_state output=nc_state_grass_1
v.proj location=nc_epsg_2264 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_grass_1
v.proj location=nc_epsg_32617 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_grass_1

reproject nc_state from nc_spm_08 to location nc_epsg_32617 to location nc_epsg_2264 and back to nc_spm_08
v.proj location=nc_spm_08 mapset=PERMANENT input=nc_state output=nc_state_grass_2
v.proj location=nc_epsg_32617 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_grass_2
v.proj location=nc_epsg_2264 mapset=PERMANENT input=nc_state_srid_1 output=nc_state_grass_2

end results in nc_spm_08 should be identical (within floating point precision limits)

This is not yet working properly because the PROJ area of use ([proj_area_set_bbox()](https://github.com/OSGeo/PROJ/blob/master/src/4D_api.cpp#L891), [proj_create_crs_to_crs()](https://github.com/OSGeo/PROJ/blob/master/src/4D_api.cpp#L1151)) is not yet set properly. The new PROJ API needs an area of use, similar to the GRASS current region, to select the best coordinate transformation method. If this area of use is not set correctly, a wrong coordinate transformation method might be selected, and reprojection can fail. Therefore this PR is WIP.

### CODE REVIEW

**Comment by marisn** [comment]
This is huge! Tested with importing vectors in EPSG:3059 without reprojecting and also with reprojecting on import into EPSG:4326 – works fine. Didn't test places where improper axis ordering might pop-up. At least in importing and displaying everything is fine.

**Comment by neteler** [comment]
Huge work @metzm ! 

The CI shows two errors with Ubuntu 16 (old but maybe still fixable):

```
make[4]: Entering directory '/home/runner/work/grass/grass/raster/r.external'
test -d OBJ.x86_64-pc-linux-gnu || mkdir -p OBJ.x86_64-pc-linux-gnu
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.external\" -o OBJ.x86_64-pc-linux-gnu/colors.o -c colors.c
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.external\" -o OBJ.x86_64-pc-linux-gnu/link.o -c link.c
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.external\" -o OBJ.x86_64-pc-linux-gnu/list.o -c list.c
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.external\" -o OBJ.x86_64-pc-linux-gnu/main.o -c main.c
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.external\" -o OBJ.x86_64-pc-linux-gnu/proj.o -c proj.c
proj.c: In function ‘check_projection’:
proj.c:76:9: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
     wkt = GDALGetProjectionRef(hDS);
         ^
proj.c:117:5: error: expected ‘}’ before ‘else’
     else {
     ^
../../include/Make/Compile.make:32: recipe for target 'OBJ.x86_64-pc-linux-gnu/proj.o' failed
make[4]: Leaving directory '/home/runner/work/grass/grass/raster/r.external'
```

and

```
make[4]: Entering directory '/home/runner/work/grass/grass/raster/r.in.gdal'
test -d OBJ.x86_64-pc-linux-gnu || mkdir -p OBJ.x86_64-pc-linux-gnu
gcc  -g -O2   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grassmods"\"   -I/usr/include/gdal -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"raster/r.in.gdal\" -o OBJ.x86_64-pc-linux-gnu/main.o -c main.c
main.c: In function ‘main’:
main.c:957:11: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
   gdalwkt = GDALGetGCPProjection(hDS);
           ^
main.c:958:10: error: too few arguments to function ‘OSRImportFromWkt’
   hSRS = OSRImportFromWkt(&gdalwkt);
          ^
In file included from /home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include/grass/gprojects.h:34:0,
                 from main.c:29:
/usr/include/gdal/ogr_srs_api.h:324:16: note: declared here
 OGRErr CPL_DLL OSRImportFromWkt( OGRSpatialReferenceH, char ** );
                ^
make[4]: *** [OBJ.x86_64-pc-linux-gnu/main.o] Error 1
../../include/Make/Compile.make:32: recipe for target 'OBJ.x86_64-pc-linux-gnu/main.o' failed
make[4]: Leaving directory '/home/runner/work/grass/grass/raster/r.in.gdal'
```

Oh, it is an antique GDAL: `Unpacking libgdal1i (1.11.3+dfsg-3build2) ...` so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?

**Comment by marisn** [comment]
> Oh, it is an antique GDAL: `Unpacking libgdal1i (1.11.3+dfsg-3build2) ...` so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?

+1 for dropping Ubuntu 16

**Comment by nilason** [comment]
> Oh, it is an antique GDAL: Unpacking libgdal1i (1.11.3+dfsg-3build2) ... so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?

Note that also ubuntu-18.04 fails. There is a compiler macro condition block in `raster/r.external/proj.c:26-116`:
```
#if GDAL_VERSION_NUM >= 3000000`)
[...]
#else
[...]
#endif
```
which causes failure with GDAL < 3.

File: raster/r.external/proj.c [hunk 1]
```diff
@@ -57,18 +104,16 @@ void check_projection(struct Cell_head *cellhd, GDALDatasetH hDS,
...
-	    if (authname && *authname && strcmp(authname, "EPSG") == 0) {
+	    if (authname && *authname) {
 		authcode = OSRGetAuthorityCode(hSRS, authkey);
 		if (authcode && *authcode) {
-		    G_debug(1, "found EPSG:%s", authcode);
-		    proj_epsg = G_create_key_value();
-		    G_set_key_value("epsg", authcode, proj_epsg);
-		}
+		    G_asprintf(&srid, "%s:%s", authname, authcode);
 	    }
```
Comment (nilason) [inline raster/r.external/proj.c:111]:
```suggestion
		    G_asprintf(&srid, "%s:%s", authname, authcode);
		}
	    }
```

File: raster/r.external/proj.c [hunk 2]
```diff
@@ -57,18 +104,16 @@ void check_projection(struct Cell_head *cellhd, GDALDatasetH hDS,
...
 	}
 	if (hSRS)
 	    OSRDestroySpatialReference(hSRS);
     }
+#endif
     else {
```
Comment (nilason) [inline raster/r.external/proj.c:117]:
```suggestion
    }
    else {
```

Comment (nilason) [reply raster/r.external/proj.c:117]: `#endif` should be placed after `else` block

**Comment by nilason** [comment]
Amazing work! So far tested the first suggested test with PROJ 7.2.1 and GDAL 3.2.1. Seem to work just fine!


Compiling with clang issued the following three warnings on files you modified. They are not caused by this PR, but maybe you wouldn't mind addressing them while you're at it.

**lib/proj/do_proj.c :**

```
/usr/bin/clang  -g -pipe -arch x86_64   -fno-common  -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include    -I/opt/local/include -DPACKAGE=\""grasslibs"\"  -I/opt/local/lib/proj7/include -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include -DRELDIR=\"lib/proj\" -o OBJ.x86_64-apple-darwin19.6.0/do_proj.o -c do_proj.c
do_proj.c:176:13: warning: enumeration values 'PJ_TYPE_TEMPORAL_DATUM', 'PJ_TYPE_ENGINEERING_DATUM', and
      'PJ_TYPE_PARAMETRIC_DATUM' not handled in switch [-Wswitch]
    switch (proj_get_type(pj)) {
            ^
do_proj.c:176:13: note: add missing switch cases
    switch (proj_get_type(pj)) {
            ^
1 warning generated.
```


**raster/r.in.gdal/main.c :**
```
/usr/bin/clang  -g -pipe -arch x86_64    -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include  -DPACKAGE=\""grassmods"\"  -I/opt/local/lib/proj7/include -I/opt/local/include  -DRELDIR=\"raster/r.in.gdal\" -o OBJ.x86_64-apple-darwin19.6.0/main.o -c main.c
main.c:1282:6: warning: ignoring return value of function declared with 'warn_unused_result' attribute
      [-Wunused-result]
            GDALRasterIO(hBand, GF_Read, 0, rowmap[row], ncols_gdal, 1,
            ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.c:1337:6: warning: ignoring return value of function declared with 'warn_unused_result' attribute
      [-Wunused-result]
            GDALRasterIO(hBand, GF_Read, 0, rowmap[row], ncols_gdal, 1,
            ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.c:1411:6: warning: ignoring return value of function declared with 'warn_unused_result' attribute
      [-Wunused-result]
            GDALRasterIO(hBand, GF_Read, 0, rowmap[row], ncols_gdal, 1,
            ^~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 warnings generated.
```



**vector/v.out.ogr/main.c :**
```
/usr/bin/clang  -g -pipe -arch x86_64    -I/Dev/grass/dist.x86_64-apple-darwin19.6.0/include -I/opt/local/include -I/opt/local/include -DPACKAGE=\""grassmods"\"  -I/opt/local/include/postgresql12 -I/opt/local/lib/proj7/include  -DRELDIR=\"vector/v.out.ogr\" -o OBJ.x86_64-apple-darwin19.6.0/main.o -c main.c
main.c:834:2: warning: ignoring return value of function declared with 'warn_unused_result' attribute
      [-Wunused-result]
        OGR_L_StartTransaction(Ogr_layer);
        ^~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~
main.c:895:2: warning: ignoring return value of function declared with 'warn_unused_result' attribute
      [-Wunused-result]
        OGR_L_CommitTransaction(Ogr_layer);
        ^~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~
2 warnings generated.
```

File: raster/r.in.gdal/main.c [hunk 3]
```diff
@@ -946,13 +946,26 @@ int main(int argc, char *argv[])
...
+		char *gdalsrid = NULL, *gdalwkt = NULL;
+		OGRSpatialReferenceH hSRS = NULL;
+
+		/* GDAL >= 3 */
+#if GDAL_VERSION_MAJOR >= 3
+		hSRS = GDALGetGCPSpatialRef(hDS);
+		const char **papszOptions = NULL;
+#else
+		gdalwkt = GDALGetGCPProjection(hDS);
+		hSRS = OSRImportFromWkt(&gdalwkt);
```
Comment (nilason) [inline raster/r.in.gdal/main.c:958]: `OGRErr OSRImportFromWkt(OGRSpatialReferenceH, char**)`

https://gdal.org/api/ogr_srs_api.html?highlight=osrimportfromwkt#_CPPv416OSRImportFromWkt20OGRSpatialReferenceHPPc

**Comment by nilason** [comment]
> so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?

I don't think that would be necessary, at least not for this enhancement.

**Comment by nilason** [comment]
I encountered problems with projecting back and forth between epsg 23700 and 3035.


```sh
#!/usr/bin/env bash

GISDBASE=$HOME/grassdata
GRASS=`which grass79`

cat << EOF > script.sh
#!/usr/bin/env bash

v.proj location=projtest_3035 mapset=PERMANENT input=point_23700_2 output=point_23700_2

echo "original:"
v.out.ascii input=point_23700@PERMANENT format=point
echo "back-and-forth:"
v.out.ascii input=point_23700_2@PERMANENT format=point
EOF

chmod +x run.sh

cat << EOF > point_23700.geojson
{
"type": "FeatureCollection",
"name": "point_23700",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:EPSG::23700" } },
"features": [
{ "type": "Feature", "properties": { }, "geometry": { "type": "Point", "coordinates": [ 737000.0, 240000.0 ] } }
]
}
EOF

$GRASS -c --text EPSG:23700:3 $GISDBASE/projtest_23700 --exec v.import input=point_23700.geojson

$GRASS -c EPSG:3035 $GISDBASE/projtest_3035 --exec v.proj location=projtest_23700 mapset=PERMANENT input=point_23700 output=point_23700_2

$GRASS --text $GISDBASE/projtest_23700/PERMANENT --exec $(pwd)/script.sh
```

results in:
```
Executing <v.import input=point_23700.geojson> ...
WARNING: Datum <Hungarian_Datum_1972> not recognised by GRASS and no
         parameters found
```
and
```
Executing <v.proj location=projtest_23700 mapset=PERMANENT input=point_23700 output=point_23700_2> ...
ERROR: North must be larger than South, but 240000 (north) <= 240000 (south
Execution of <v.proj location=projtest_23700 mapset=PERMANENT input=point_23700 output=point_23700_2> finished.
```


While using ogr2ogr is pretty straightforward:

```
ogr2ogr -t_srs EPSG:3035 -f GeoJSON point_3035.geojson point_23700.geojson
ogr2ogr -t_srs EPSG:23700 -f GeoJSON point_23700_2.geojson point_3035.geojson
```

**Comment by wenzeslaus** [comment]
> ...so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?

Fine with me. Fair enough for 8.0 and a major improvement like this one.

File: raster/r.external/proj.c (hunk 2 above)
Comment (metzm) [reply raster/r.external/proj.c:118]: No, the `#endif` must be before the `else` block. Above, there are two different `if` blocks, one for GDAL >= 3, another one for GDAL < 3.

**Comment by metzm** [comment]
Thanks all for your helpful comments!

I have fixed support for older GDAL/PROJ versions. Removing support for older GDAL/PROJ versions would be very helpful! The code base of the respective libraries and modules could be substantially cleaned up (removing lots of these `#if*` blocks. But this would be another big PR.

Compiler warnings have been taken care of, and reprojection of a single point works now.

As long as my original test fails, this PR will remain a draft. I am looking into it.

**Comment by neteler** [comment]
> > ...so perhaps we just drop Ubuntu 16 from GitHub actions (@wenzeslaus , others)?
> 
> Fine with me. Fair enough for 8.0 and a major improvement like this one.

Ok, see #1254

**Comment by metzm** [comment]
The last commit [ab7a768](https://github.com/OSGeo/grass/pull/1240/commits/ab7a768430c9405539e1b77e06b728006b59ecae) fixes issues with the area of interest to be set for PROJ 6+. However, my tests still fail.
Test case: create a location with EPSG:2264, then try `r.proj loc=nc_spm_08 map=PERMANENT in=landcover_1m -g` or `r.proj loc=nc_spm_08 map=PERMANENT in=elev_state_500m`
At least `r.proj loc=nc_spm_08 map=PERMANENT in=elev_state_500m -g` is working...

**Comment by metzm** [comment]
There is a bug in PROJ causing my tests to fail with `point not within available datum shift grids`. The bug has been fixed in PROJ master with [ebe3425](https://github.com/OSGeo/PROJ/commit/ebe3425bf66287e004958eb53976d3837f88b9e1). Still, I have to rewrite the relevant code in GRASS lib/proj to avoid this bug in PROJ.

**Comment by metzm** [comment]
This PR (together with previous commits) would fix bugs also present in G78: 

[CRS mismatch](https://github.com/qgis/QGIS/issues/18596#issuecomment-753481210) between input and output when creating a location from some GDAL/OGR recognized input, importing to this location and exporting again. The CRS should be exactly preserved. This bug happens when GRASS is compiled with PROJ6+ and GDAL3.

A bug in PROJ can cause PROJ to select a wrong coordinate operation with a datum transformation grid that does not cover the whole area of interest. In these cases, reprojection fails. This PR takes care of this bug.

Fixing these bugs in G78 requires that all changes related to the new SRID and WKT2 handling including all affected import and export modules need to be backported.

**Comment by nilason** [comment]
My previous observations, issues have all been solved.
Plan to do some more testing, but I saw an error while reprojecting in GUI Data tab:
 
```
r.proj dbase=/grassdata location=nc_spm_08 mapset=user1 input=testelev output=testelev method=bilinear
TypeError
:
Validate() takes 1 positional argument but 2 were given
```
This may be an issue for a follow-up PR though.

File: general/g.proj/input.c [hunk 4]
```diff
@@ -372,26 +372,110 @@ int input_epsg(int epsg_num)
...
+	const char **papszOptions = NULL;
+
+	ret = GPJ_osr_to_grass(&cellhd, &projinfo, &projunits, hSRS, 0);
+
+	if (cellhd.proj == PROJECTION_XY)
+	    G_warning(_("Read of file %s was successful, but it did not contain "
+			"projection information. 'XY (unprojected)' will be used"),
+		      geofile);
+
+	papszOptions = G_calloc(3, sizeof(char *));
```
Comment (rouault) [inline general/g.proj/input.c:465]: this is not freed

File: raster/r.out.gdal/main.c [hunk 5]
```diff
@@ -286,7 +286,69 @@ int main(int argc, char *argv[])
...
+	    }
+	    proj_destroy(obj);
+	}
+    }
+    if (!srswkt && (indef = G_get_projwkt())) {
+	OGRSpatialReferenceH hSRS;
+	const char **papszOptions = NULL;
+
+	hSRS = OSRNewSpatialReference(indef);
+	papszOptions = G_calloc(2, sizeof(char *));
```
Comment (rouault) [inline raster/r.out.gdal/main.c:312]: this is not freed

File: vector/v.out.ogr/main.c [hunk 6]
```diff
@@ -267,7 +267,64 @@ int main(int argc, char *argv[])
...
+
+#if GDAL_VERSION_MAJOR >= 3 && PROJ_VERSION_MAJOR >= 6
+	if (Ogr_projection) {
+	    /* convert bound CRS */
+	    PJ *obj = NULL;
+	    const char *papszOptions[2];
+
+	    inwkt = NULL;
+
+	    papszOptions[0] = G_store("FORMAT=WKT2");
```
Comment (rouault) [inline vector/v.out.ogr/main.c:304]: this is not freed

File: vector/v.out.ogr/main.c [hunk 7]
```diff
@@ -267,7 +267,64 @@ int main(int argc, char *argv[])
...
+	    papszOptions[1] = NULL;
+	    OSRExportToWktEx(Ogr_projection, &inwkt, papszOptions);
```
Comment (rouault) [inline vector/v.out.ogr/main.c:306]: inwkt should be freed with CPLFree()

File: vector/v.out.ogr/main.c [hunk 8]
```diff
@@ -267,7 +267,64 @@ int main(int argc, char *argv[])
...
+		    G_debug(1, "found bound crs");
+		    source_crs = proj_get_source_crs(NULL, obj);
+		    if (source_crs) {
+			inwkt = G_store(proj_as_wkt(NULL, source_crs, PJ_WKT2_2019, NULL));
+			if (inwkt && !*inwkt) {
+			    G_free(inwkt);
+			    inwkt = NULL;
+			}
+			proj_destroy(source_crs);
+			OSRImportFromWkt(Ogr_projection, &inwkt);
```
Comment (rouault) [inline vector/v.out.ogr/main.c:321]: use
```
char* inwktTmp =inwkt;
OSRImportFromWkt(Ogr_projection, &inwktTmp);
```
to avoid inwkt to be modified

**Comment by metzm** [comment]
> My previous observations, issues have all been solved.
> Plan to do some more testing, but I saw an error while reprojecting in GUI Data tab:
> 
> ```
> r.proj dbase=/grassdata location=nc_spm_08 mapset=user1 input=testelev output=testelev method=bilinear
> TypeError
> :
> Validate() takes 1 positional argument but 2 were given
> ```
> 
> This may be an issue for a follow-up PR though.

This looks like a GUI issue. Please provide the full GUI output.

**Comment by metzm** [comment]
> My previous observations, issues have all been solved.
> Plan to do some more testing, but I saw an error while reprojecting in GUI Data tab:
> 
> ```
> r.proj dbase=/grassdata location=nc_spm_08 mapset=user1 input=testelev output=testelev method=bilinear
> TypeError
> :
> Validate() takes 1 positional argument but 2 were given
> ```
> 
> This may be an issue for a follow-up PR though.

Not reproducible through the menu "Raster -> Develop raster map -> Reproject raster map from different location".

Must be a bug in the Data tab.

File: general/g.proj/input.c (hunk 4 above)
Comment (metzm) [reply general/g.proj/input.c:471]: Done.

File: raster/r.out.gdal/main.c (hunk 5 above)
Comment (metzm) [reply raster/r.out.gdal/main.c:312]: Done.

File: vector/v.out.ogr/main.c (hunk 6 above)
Comment (metzm) [reply vector/v.out.ogr/main.c:304]: Done.

File: vector/v.out.ogr/main.c (hunk 7 above)
Comment (metzm) [reply vector/v.out.ogr/main.c]: Done.

File: vector/v.out.ogr/main.c (hunk 8 above)
Comment (metzm) [reply vector/v.out.ogr/main.c]: Done.

**Comment by metzm** [comment]
Thanks very much @rouault for your suggestions!
I think this PR is now a working solution with reasonably good code quality and no apparent errors. As always, the new code can always be improved.

**Comment by nilason** [comment]
> Not reproducible through the menu "Raster -> Develop raster map -> Reproject raster map from different location".
>
> Must be a bug in the Data tab.

Yes, its only in Data tab and also unrelated to this PR. Reported with #1273.

**Comment by mlennert** [comment]
> Thanks very much @rouault for your suggestions!
> I think this PR is now a working solution with reasonably good code quality and no apparent errors. As always, the new code can always be improved.

No possibility for me to test this seriously in the near future, but big kudos to @metzm for this great improvement !

**Comment by metzm** [comment]
@neteler backported to G78

**Pull request merged**